### PR TITLE
[#4608] feat(core): Supports to load fields lazily

### PR DESCRIPTION
**Pull request by jerqi**
### What changes were proposed in this pull request?

Supports to load fields lazily.

### Why are the changes needed?

Fix: #4608

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Add new uts and pass the existing uts

### CODE REVIEW

**Comment by jerryshao** [comment]
I will wait this after the 0.6.0 release.

**Comment by jerqi** [comment]
@xunliu @yuqi1129 @jerryshao Could you take a look?

File: core/src/main/java/org/apache/gravitino/meta/GroupEntity.java [hunk 1]
```diff
@@ -38,21 +39,25 @@ public class GroupEntity implements Group, Entity, Auditable, HasIdentifier {
...
   private Long id;
   private String name;
   private AuditInfo auditInfo;
-  private List<String> roleNames;
-  private List<Long> roleIds;
   private Namespace namespace;
+  // The roleIds is a lazy field to avoid unnecessary cost.
+  private Supplier<List<Long>> roleIdsSupplier = () -> null;
+  // The roleNames is a lazy field to avoid unnecessary cost.
+  private Supplier<List<String>> roleNamesSupplier = () -> null;
```
Comment (xunliu) [inline core/src/main/java/org/apache/gravitino/meta/GroupEntity.java:60]: I think we need to use one `Role` object variable to replace `roleIds` and `roleNames` two variable.

File: core/src/main/java/org/apache/gravitino/meta/UserEntity.java [hunk 2]
```diff
@@ -44,18 +45,22 @@ public class UserEntity implements User, Entity, Auditable, HasIdentifier {
...
   private Long id;
   private String name;
   private AuditInfo auditInfo;
-  private List<String> roleNames;
-  private List<Long> roleIds;
   private Namespace namespace;
+  // The roleIds is a lazy field to avoid unnecessary cost.
+  private Supplier<List<Long>> roleIdsSupplier = () -> null;
+  // The roleNames is a lazy field to avoid unnecessary cost.
+  private Supplier<List<String>> roleNamesSupplier = () -> null;
```
Comment (xunliu) [inline core/src/main/java/org/apache/gravitino/meta/UserEntity.java:63]: I think we need to use one `Role` object variable to replace `roleIds` and `roleNames` two variable.

File: core/src/main/java/org/apache/gravitino/storage/relational/utils/SupplierUtils.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  public static Supplier<List<SecurableObject>> createSecurableObjectsSupplier(RolePO rolePO) {
+    return new Supplier<List<SecurableObject>>() {
+      private List<SecurableObject> securableObjects;
+
+      private boolean waitToLoad = true;
+
+      @Override
+      public List<SecurableObject> get() {
+        if (waitToLoad) {
+          waitToLoad = false;
```
Comment (yuqi1129) [inline core/src/main/java/org/apache/gravitino/storage/relational/utils/SupplierUtils.java:85]: The value of `waitToLoad` should be set to false after `securableObjects` has been assigned successfully.

File: core/src/main/java/org/apache/gravitino/meta/GroupEntity.java (hunk 1 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/meta/GroupEntity.java]: Done.

File: core/src/main/java/org/apache/gravitino/storage/relational/utils/SupplierUtils.java (hunk 3 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/storage/relational/utils/SupplierUtils.java]: Done.

File: core/src/main/java/org/apache/gravitino/storage/relational/utils/SupplierUtils.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+              RoleMetaService.getInstance().listSecurableObjectsByRoleId(rolePO.getRoleId());
+
+          securableObjects = Lists.newArrayList();
+
+          for (SecurableObjectPO securableObjectPO : securableObjectPOs) {
+            String fullName =
+                MetadataObjectService.getMetadataObjectFullName(
+                    securableObjectPO.getType(), securableObjectPO.getMetadataObjectId());
+            if (fullName != null) {
+              securableObjects.add(
```
Comment (yuqi1129) [inline core/src/main/java/org/apache/gravitino/storage/relational/utils/SupplierUtils.java:95]: I believe you need to clear `securableObjects` before you try to load them from storage, `securableObjects` is not empty when `get` fails halfway.

Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/storage/relational/utils/SupplierUtils.java:95]: You can see the line 88. I have cleared it.

File: core/src/main/java/org/apache/gravitino/meta/RoleEntity.java [hunk 5]
```diff
@@ -91,16 +95,7 @@ public Map<String, String> properties() {
...
   public List<SecurableObject> securableObjects() {
-    // The securable object is a special kind of entities. Some entity types aren't the securable
-    // object, such as
-    // User, Role, etc.
-    // The securable object identifier must be unique.
-    // Gravitino assumes that the identifiers of the entities may be the same if they have different
-    // types.
-    // So one type of them can't be the securable object at least if there are the two same
-    // identifier
-    // entities .
```
Comment (xunliu) [inline core/src/main/java/org/apache/gravitino/meta/RoleEntity.java:102]: We don't need these comments any more?

Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/meta/RoleEntity.java:102]: This is legacy comments. The interface has more accurate comments.

**Review by xunliu — APPROVED** [review]
LGTM

**Pull request merged**